src/renderer/viz/expressions/viewportFeatures.js

```diff
@@ -84,34 +87,31 @@ export default class ViewportFeatures extends BaseExpression {
     }
 
     _resetViewportAgg () {
-        if (!this._FeatureProxy) {
+        if (!this._columns) {
             if (!this._requiredProperties.every(p => (p.isA(Property)))) {
                 throw new CartoValidationError(`${cvt.INCORRECT_TYPE} viewportFeatures arguments can only be properties`);
             }
+
             const columns = Object.keys(schema.simplify(this._getMinimumNeededSchema()));
-            this._FeatureProxy = this.genViewportFeatureClass(columns);
+            FEATURE_VIZ_PROPERTIES.forEach((propertyName) => {
+                if (columns.includes(propertyName)) {
+                    throw new CartoValidationError(`${cvt.INCORRECT_VALUE} '${propertyName}' property can't be used, as it is a reserved property name`);
+                }
+            });
+            this._columns = columns;
         }
         this.expr = [];
     }
 
-    accumViewportAgg (feature) {
-        this.expr.push(new this._FeatureProxy(feature));
+    accumViewportAgg (interactivityFeature) {
+        this._addRequiredPropertiesTo(interactivityFeature);
+        this.expr.push(interactivityFeature);
     }
 
-    genViewportFeatureClass (properties) {
-        const cls = class ViewportFeature {
-            constructor (feature) {
-                this._feature = feature;
-            }
-        };
-        properties.forEach(prop => {
-            Object.defineProperty(cls.prototype, prop, {
-                get: function () {
-                    return this._feature[prop];
-                }
-            });
+    _addRequiredPropertiesTo (interactivityFeature) {
+        this._columns.forEach((column) => {
+            interactivityFeature[column] = interactivityFeature.rawFeature[column];
```

Comment:
This refactor is not the first time it is proposed. However, there is a reason for `genViewportFeatureClass`. 

`genViewportFeatureClass` defines a class with getters which are able to retrieve the properties without storing the properties within the class instance.

The previous code, which is similar to yours, adds a lot of overhead in memory allocation and in GC since each feature is being created one property at a time.

See the PR that optimized this: https://github.com/CartoDB/carto-vl/pull/663

Reply:
Fair enough, thx!

Reply:
I have applied the same dynamic creation of getters, but not with a proxy in the class, but within the Feature class itself, with an optional parameter. See ⬇️